Relay compact block messages prior to full block connection

TheBlueMatt opened the pull request:
First clean up net_processing to never have non-const access to CBlockIndex*es.

Then take advantage of #9026 to relay compact block announcements prior to full validation of the block. This has the (incredibly) nice property of not re-deserializing the same block over and over again in SendMessages when we go to announce the block.

Finally caches the block which was just announced to respond to getblocktxn requests without a similar deserialize (this could probably be extended to be used by other requests as well, but other requests are still going to want to lock cs_main and check that the block didn't get rejected as invalid first).

TheBlueMatt committed:
Cache most-recently-announced block's shared_ptr

TheBlueMatt committed:
Make CBlockIndex*es in net_processing const

TheBlueMatt committed:
Call AcceptBlock with the block's shared_ptr instead of CBlock&

TheBlueMatt committed:
Add a CValidationInterface::NewPoWValidBlock callback

gmaxwell commented:
could we have some benchmarks do demonstrate the gains from this? though I don't doubt them, this is a fair amount of code...  and benchmarks would be important for regression testing this important behavior.

TheBlueMatt commented:
Which part, specifically, would you like to see benchmarks on?

On December 18, 2016 11:40:22 PM PST, Gregory Maxwell <notifications@github.com> wrote:
>could we have some benchmarks do demonstrate the gains from this?
>though I don't doubt them, this is a fair amount of code...  and
>benchmarks would be important for regression testing this important
>behavior.
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/bitcoin/bitcoin/pull/9375#issuecomment-267900573

gmaxwell commented:
latency to relay a block across many connections would be the obvious thing?

TheBlueMatt committed:
[qa] Avoid race in preciousblock test.

If node 0 is sufficiently fast to announce its block to node 1,
node 1 might already have the block by the time the
node_sync_via_rpc loop gets around to node 1, resulting in the
submitblock result "duplicate-inconclusive" as node 1 has the block,
but prefers an alternate chain.

morcos commented:
Is there ever a case where we would accept a block ourselves but we wouldn't want to immediately announce it?  I looked through all the places we call AcceptBlock/ProcessNewBlock and couldn't immediately see a problem, but want to make sure we think about it carefully.  I noticed this because I found it odd you had to change the preciousblock.py test.   BTW, a better change there may just be to allow duplicate-inconclusive in the list of results..

rebroad commented on src/net_processing.cpp:
```diff
@@ -1492,6 +1525,26 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
+                for (size_t i = 0; i < req.indexes.size(); i++) {
+                    if (req.indexes[i] >= block.vtx.size()) {
+                        Misbehaving(pfrom->GetId(), 100);
+                        LogPrintf("Peer %d sent us a getblocktxn with out-of-bounds tx indices", pfrom->id);
```
this chunk of code is duplicated elsewhere?

TheBlueMatt replied:
OK, DRY'd.

morcos commented on src/net_processing.cpp:
```diff
@@ -752,6 +752,29 @@ void PeerLogicValidation::SyncTransaction(const CTransaction& tx, const CBlockIn
+        // but we don't think they have this one, go ahead and announce it
+        if (state.fPreferHeaderAndIDs && (!fWitnessEnabled || state.fWantsCmpctWitness) &&
+                !PeerHasHeader(&state, pindex) && PeerHasHeader(&state, pindex->pprev)) {
+            connman->PushMessage(pnode, msgMaker.Make(NetMsgType::CMPCTBLOCK, cmpctblock));
```
maybe add ` LogPrint("net", "%s sending header-and-ids %s to peer %d..`

TheBlueMatt replied:
Fixed.

morcos commented on src/net_processing.cpp:
```diff
@@ -752,6 +752,39 @@ void PeerLogicValidation::SyncTransaction(const CTransaction& tx, const CBlockIn
+    CNetMsgMaker msgMaker(PROTOCOL_VERSION);
+
+    LOCK(cs_main);
+    bool fWitnessEnabled = IsWitnessEnabled(pindex, Params().GetConsensus());
```
shouldn't this be pindex->pprev?  I don't suppose it matters much, but I think you'll be thinking segwit activated 1 block early

TheBlueMatt replied:
Fixed.

rebroad replied:
@morcos don't you mean one block late?

morcos commented:
This seems like its going to be super awesome.
code review ACK modulo nits, doing some testing...

TheBlueMatt commented:
Fixed comments, note that the SendBlockTransactions function seems a bit odd now, but the goal longer-term is to move messages into their own processing groups, where such functions will be really sane.

gmaxwell commented:
Perhaps Newpowvalidblock should be gated on there actually being any peers that request a compact block from us?  Otherwise we'll go through the work of constructing a compact block there only to send it to no one. :)

Thoughts on caching the compact block like you cache the block?  You would want it for non-pref peers that request it after you announce the header... and it's reasonably small.

morcos commented:
utACK 5bc61fa

TheBlueMatt committed:
Cache most-recently-connected compact block

TheBlueMatt commented:
@gmaxwell, well those requests are somewhat contradictory, so I picked the cache-compact-representation option :p.

gmaxwell commented:
@TheBlueMatt not so! as it could skip processing when there are no peers requesting, and update caches any time it processes!

I don't think it matters except in one case:

Isolated mining node with no HB requesting peers, constructing the short block earlier will delay block validation which delays updating the create new block.  I'm guessing the delay is <10ms, so probably inconsequential. But thought I'd bring it to your attention.

TheBlueMatt commented:
Yea, I'm not worried about the processing time for calling that once. Long term I'd like to run most of the CValidationInterface callbacks into background threads, but that's probably an 0.15-targeted thing.

On December 23, 2016 1:08:51 PM EST, Gregory Maxwell <notifications@github.com> wrote:
>@TheBlueMatt not so! as it could skip processing when there are no
>peers requesting, and update caches any time it processes!
>
>I don't think it matters except in one case:
>
>Isolated mining node with no HB requesting peers, constructing the
>short block earlier will delay block validation which delays updating
>the create new block.  I'm guessing the delay is <10ms, so probably
>inconsequential. But thought I'd bring it to your attention.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/bitcoin/bitcoin/pull/9375#issuecomment-269026347

gmaxwell commented on src/net_processing.cpp:
```diff
@@ -752,6 +752,29 @@ void PeerLogicValidation::SyncTransaction(const CTransaction& tx, const CBlockIn
+
+    connman->ForEachNode([this, &cmpctblock, pindex, &msgMaker, fWitnessEnabled](CNode* pnode) {
+        // TODO: Avoid the repeated-serialization here
+        if (pnode->nVersion < INVALID_CB_NO_BAN_VERSION)
```
I believe this needs a fDisconnect check.

TheBlueMatt replied:
Fixed

TheBlueMatt commented:
Squashed with no diff-tree.

theuni commented on src/validation.cpp:
```diff
@@ -3029,12 +3029,14 @@ static bool AcceptBlockHeader(const CBlockHeader& block, CValidationState& state
         LOCK(cs_main);
         for (const CBlockHeader& header : headers) {
-            if (!AcceptBlockHeader(header, state, chainparams, ppindex)) {
+            // cast away the ppindex-returns-const CBlockIndex - we're just assigning it to a CBlockIndex*
```
Hmm, this is non-obvious to a caller. And it's especially bad that one of them relies on the const not actually being const :(

Maybe return an updated index, or null in the case of failure? Or return a pair?

TheBlueMatt replied:
Really? I dont see anything non-obvious here? The caller passes in a reference to a const CBlockIndex*, and the function sets that element to the new CBlockIndex*, ie it is returning a const CBlockIndex*, not editing a CBlockIndex which was passed in.

ryanofsky replied:
To avoid the const cast, you could do something like:

```
            CBlockIndex* pindex = nullptr;
            if (!AcceptBlockHeader(header, state, chainparams, &pindex)) {
                return false;
            }
            *ppindex = pindex;
```

TheBlueMatt replied:
OK, changed.

theuni commented on src/validation.cpp:
```diff
@@ -3799,16 +3802,19 @@ bool LoadExternalBlockFile(const CChainParams& chainparams, FILE* fileIn, CDiskB
                     std::pair<std::multimap<uint256, CDiskBlockPos>::iterator, std::multimap<uint256, CDiskBlockPos>::iterator> range = mapBlocksUnknownParent.equal_range(head);
                     while (range.first != range.second) {
                         std::multimap<uint256, CDiskBlockPos>::iterator it = range.first;
-                        if (ReadBlockFromDisk(block, it->second, chainparams.GetConsensus()))
```
Maybe teach ReadBlockFromDisk to read to a shared_ptr reference to avoid the temptation?

theuni replied:
Or just use a new precursiveblock here with limited scope

TheBlueMatt replied:
Fixed.

theuni commented on src/net_processing.cpp:
```diff
@@ -752,6 +752,33 @@ void PeerLogicValidation::SyncTransaction(const CTransaction& tx, const CBlockIn
+
+    LOCK(cs_main);
+    bool fWitnessEnabled = IsWitnessEnabled(pindex->pprev, Params().GetConsensus());
+    uint256 hashBlock(pblock->GetHash());
```
How about generating a set of to-be-sent nodes here, with cs_main scope-locked. Then pass the set into ForEachNode and send only to the matches?

That saves cs_main (or whatever lock) from being held for the duration, and opens the door to deduped sending.

TheBlueMatt replied:
AFICT the only difference is that we can skip calling connman->PushMessage....is connman->PushMessage slow enough to care?

theuni replied:
You can avoid calling it for each node by iterating mapNodeState for the set up front.

One PushMessage is no concern, but avoiding it for the entire ForEachNode would be nice.

theuni replied:
Actually, nevermind. We can improve CConnman's interface for this kind of thing later.

TheBlueMatt replied:
Yup, that was my (too heavily implied) point :).

theuni commented on src/net_processing.cpp:
```diff
@@ -760,6 +764,12 @@ void PeerLogicValidation::NewPoWValidBlock(const CBlockIndex *pindex, const std:
     bool fWitnessEnabled = IsWitnessEnabled(pindex->pprev, Params().GetConsensus());
     uint256 hashBlock(pblock->GetHash());
 
+    {
```
Make this a member of PeerLogicValidation, then you can use a weak_ptr for the global, and I think no locks are needed

TheBlueMatt replied:
I'm confused as to exactly what you're suggesting here. I don't particularly want a weak_ptr because I do want to hold the latest block for responses for a while after the block is connected/broadcast.

theuni replied:
These are the types of changes that make me nervous about shared_ptrs, because it gets easy to start stashing things that have no obvious lifetime constraint.

I was suggesting that you hold the shared_ptr in the PeerLogicValidation class. You then use a global weak_ptr which is overwritten by each new block. Then at any random time you can use std::shared_ptr\<const CBlock\> bestblock = shared_ptrglobal_last_block.lock().

That way you can extend the lifetime of the whatever block is current, but it's also obvious when it will be destroyed (with the PeerLogicValidation destruction).

TheBlueMatt replied:
Hmm...does this make it that much more obvious? We still have a shared_ptr which is overwritten in NewPoWValidBlock and other sites which take references to it to extend its lifetime. I can add a comment noting that users MUST NOT extend the lifetime significantly, would that be sufficient to address your concern?

theuni commented on src/net_processing.cpp:
```diff
@@ -1082,6 +1092,23 @@ uint32_t GetFetchFlags(CNode* pfrom, const CBlockIndex* pprev, const Consensus::
+    }
+    LOCK(cs_main);
+    CNetMsgMaker msgMaker(pfrom->GetSendVersion());
+    int nSendFlags = State(pfrom->GetId())->fWantsCmpctWitness ? 0 : SERIALIZE_TRANSACTION_NO_WITNESS;
```
No need to stay locked for PushMessage

TheBlueMatt replied:
Is it slow enough to matter? After #9414 (and some std::atomic usage) we need to remove the cs_main lock completely here to make getblocktxn for the current head cs_main-free.

TheBlueMatt commented:
Fixed one of @theuni's comments and rebased. Diff is https://0bin.net/paste/lZfeK4vg0ZdebkqX#AJKsFyZ3Ai1oiaRiUt-1S0tjFK9wrZ70aP74IepwbsY

gmaxwell commented:
One behavior that should be mentioned is that caching means that if I am connected to a since node via multiple channels, it will identify itself as the same node by using the same salt across them.  The salt sharing is something the protocol was carefully designed to enable, and there are _many_ other ways that a node identifies itself as equal on different links: so I don't consider it a concern. But since we do sometimes fix 'fingerprinting' it might surprise someone.

(My view is that we fix fingerprinting that happens at different times-- that the host that was at 1.2.3.4 is now at 2.3.4.5--, but verifying that two peers you are currently connected to are really the same peer, is too hard to avoid.)

morcos commented:
re-ACK 8022035

huge improvement to cache the cmpctblock, reduces time to relay cmpctblock to each additional peer from 30ms to 1ms.

morcos commented:
Mentioned on IRC, combining this with #9400 can lead to intermittent failures of `sendheaders.py`

#9400 causes the node announcing the reorg in that test to be a HB peer, and block requests generated by FindNextBlocksToDownload in response to cmpctblocks relayed by NewPoWValidBlock are in a race condition against the tip being updated in the announcing peer.

TheBlueMatt commented:
As @morcos points out, this needs a bit more thought - currently you might announce a compact block in HB mode, the receiving peer might then, for various reasons, request a response in the form of a full block instead of as a getblocktxn (eg #8800). At this point, you unlock cs_main prior to ActivateBestChain in ProcessNewBlock, allowing the other peer's block request the be processed, but you will not respond because you "dont have the block". A simple fix might be calling ActivateBestChain prior to block requests (since we cannot simply respond with the block without having validated it - this would be a protocol violation).

TheBlueMatt committed:
Ensure we meet the BIP 152 old-relay-types response requirements

In order to do this, we must call ActivateBestChain prior to
responding getdata requests for blocks which we announced using
compact blocks.

For getheaders responses we dont need code changes, but do note
that we must reset the bestHeaderSent so that the SendMessages call
re-announces the header in question.

While we could do something smarter for getblocks, calling
ActivateBestChain is simple and more obviously correct, instead of
doing something more similar to getheaders.

See-also the BIP clarifications at
https://github.com/bitcoin/bips/pull/486

TheBlueMatt commented:
Pushed a new commit which appears to fix @morcos' test failures in #9447. See the BIP clarifications at https://github.com/bitcoin/bips/pull/486 for associated info.

rebroad commented on src/net_processing.cpp:
```diff
@@ -752,6 +752,45 @@ void PeerLogicValidation::SyncTransaction(const CTransaction& tx, const CBlockIn
+        if (state.fPreferHeaderAndIDs && (!fWitnessEnabled || state.fWantsCmpctWitness) &&
+                !PeerHasHeader(&state, pindex) && PeerHasHeader(&state, pindex->pprev)) {
+
+            LogPrint("net", "%s sending header-and-ids %s to peer %d\n", "PeerLogicValidation::NewPoWValidBlock",
```
`__func__` can be used here?

TheBlueMatt replied:
I'm not sure that it can due to the use of the inline function - do you care to test what __FUNC__ means on different compilers in this case?

morcos replied:
I tried that in my first patch to print this.  it printed "operator ()"

morcos commented on src/net_processing.cpp:
```diff
@@ -1507,6 +1517,18 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
+        // compact blocks but still use getblocks to request blocks.
+        {
+            CValidationState dummy;
+            ActivateBestChain(dummy, Params());
```
Just a note that we'll be calling this for every GETBLOCKS request.  But I suppose ActivateBestChain isn't too expensive in the common case.

TheBlueMatt replied:
Yea, its lightweight-enough that we cant really consider it a DoS issue (please review, but I'm pretty sure here), and its not like we care much about the performance of GETBLOCKS anymore...

theuni replied:
Mentioned on IRC as well:

An example of an unintended side-effect of this:
- miner mines a block, hands it to ProcessNewBlock()
- ProcessMessages thread simultaneously receives a GETBLOCKS
- ProcessMessages calls ActivateBestChain _just_ after miner finishes AcceptBlock() and releases cs_main.
- ConnectTip ends up getting called with a null block, activating the chain with the miner's new block, but forcing it to be read from disk
- Miner's ActivateBestChain is a no-op

The above is unlikely and harmless (minus the msec it takes to read/de-serialize from disk), but it's an example of how someone hammering a miner with GETBLOCKS would be capable of affecting validation.
Presumably with multi-threaded message processing, a node hammering with GETBLOCKS might be able to force all block reads to disk this way.

If the intention is simply to ensure that we've fully validated all known blocks at this moment, I'd be much more comfortable with some function that simply blocks until any current ActivateBestChain() completes. That would also protect us from oopses in the future when processing becomes multi-threaded.

TheBlueMatt replied:
@sipa Yes, moved this one out of the cs_main, but the other one is much harder to move out of cs_main (luckily the one that handles getdatas is much less likely to have long-runtime because its specifically gated already).

@theuni Will follow up on IRC, though I fixed the particular issue you mentioned (not that its a response to your general concern).

sipa commented:
I wish that we had a way to announce that we accept blocks that are not
fully validated (as long as they satisfy PoW and size limits). Then compact
blocks could be orthogonal to the choice of relay before/after validation.

morcos commented:
ACK 73817e2

instagibbs commented:
utACK https://github.com/bitcoin/bitcoin/pull/9375/commits/802203579d764bf24fbd3b0a5addc916ba4195e3

sdaftuar commented on src/net_processing.cpp:
```diff
@@ -2785,13 +2787,19 @@ bool SendMessages(CNode* pto, CConnman& connman)
+
+                    LOCK(cs_most_recent_block);
+                    if (most_recent_block_hash == pBestIndex->GetBlockHash()) {
+                        connman.PushMessage(pto, msgMaker.Make(nSendFlags, NetMsgType::CMPCTBLOCK, *most_recent_compact_block));
```
The cached compact block was built using wtxid's rather than txid's, so we have to check to see if the peer wants compact witness ids before we can use the cached version, no?

sdaftuar commented:
> Is there ever a case where we would accept a block ourselves but we wouldn't want to immediately announce it?

@morcos @TheBlueMatt One area worth some thought is how to handle block relay while processing a reorg.  Before this PR, we wouldn't relay any block with less work than our tip unless we had completed a reorg to a new tip.  After this PR, I believe we would relay blocks with work less than or equal to our tip (as compact blocks, even) before we know that the reorg would succeed.

And in fact, upon receipt of a compact block with <= work than our tip, we don't bother trying to process.  So I'm not sure there's any benefit to relaying before validation for blocks that aren't just building on our tip?

sdaftuar commented:
Actually I think there are some problems with relaying blocks that have work equal to our tip.  Conceptually, we shouldn't be helping our peers end up on a tip that is competing with ours, so even though we download blocks at the same work as our tip, I think it's important that we not relay such blocks.

More practically, in the CMPCTBLOCK handling code, we call `UpdateBlockAvailability()` after processing the header, which will overwrite the value of `pindexBestKnownBlock` for our peer.  I believe this means that if we relay a block that has the same work as our tip, then our peers will "forget" that they can even download our actual tip from us(!).

My suggestion: only call `NewPoWValidBlock()` in `AcceptBlock()` for blocks that build on our tip.

sdaftuar commented on src/net_processing.cpp:
```diff
@@ -949,6 +949,16 @@ void static ProcessGetData(CNode* pfrom, const Consensus::Params& consensusParam
                 BlockMap::iterator mi = mapBlockIndex.find(inv.hash);
                 if (mi != mapBlockIndex.end())
                 {
+                    if (mi->second->nChainTx && !mi->second->IsValid(BLOCK_VALID_SCRIPTS) &&
```
I should think this through a bit more, but I believe the requirement that the block be VALID_SCRIPTS is stronger than is needed -- in particular, if two blocks arrive simultaneously, we might relay both before picking which of the two is our tip, and then gating ProcessGetData() on VALID_SCRIPTS will cause us to stall a peer who picked the other.

Perhaps we should change VALID_SCRIPTS to VALID_TRANSACTIONS below, and then actually I'm not sure if we would still need this code path?

morcos replied:
@gmaxwell see this comment.  I think this would cover the case of needing to cache more than one block..

TheBlueMatt replied:
This seems like a rather large change (that I dont particularly want to do in this PR). Is it sufficient to not do the pre-forward unless the new block builds on our tip?

morcos replied:
commented on IRC, but yes i think we need a fix here.

Otherwise we can be in a situation where we stall a peer who asks for a block we just announced to them.

i.e. We try and connect A and A' near simultaneously as new tips.  We announce both, but whichever gets connected first, the other we will stall requests for.  This can happen even without multi-threaded ProcessMessages via a race with submitblock.

sdaftuar replied:
ping @sipa, please see above...

sdaftuar replied:
Ok after further thought and offline discussion, I retract my objection to leaving the VALID_SCRIPTS check in.

My original concern was that there were some race conditions where we might announce a block pre-validation, and then never test the block's validity, resulting in stalling block download to a peer for a valid block.  After the change to only do pre-validation announcements for a single block at a given height and only for blocks that build on our tip, I think this is no longer an issue.

A secondary concern was generally figuring out how to handle block requests for invalid blocks.  In the typical case of a cmpctblock announcement being followed by a getblocktxn request, I believe everything works as intended (a node would provide the blocktxn, and the peer would not punish if the block turned out to be invalid).  However if a peer somehow fell back to a getdata request for that block, then this code would cause us to stall the peer's block download, likely resulting in them disconnecting us.

The alternative, however, of delivering the block when we know it is invalid is potentially even worse -- they would ban us, absent more code changes to handle this situation (which #9026 overlooked), which is a lot more code complexity.

I haven't been able to come up with realistic scenarios where someone could provoke these conditions (ie, produce an invalid block in such a way as to also cause nodes to fall back to getdata responses to the cmpctblock announcement), so I think this is an acceptable behavior to introduce for now.  In the future, though it would be nice to add a protocol extension so that we could have the option of telling our peer that a block request is being ignored, so that the peer could do something smarter than just disconnect us for stalling download.

To remind us, I think it'd be nice to add a comment in `ProcessGetData()` that mentions this issue, something like:

```
// TODO: extend the protocol to allow us to tell our peer that we're not going to deliver a block that we've previously announced, eg because we don't think the block is valid.
// This would give allow our peer to be make a better decision than just wait indefinitely or disconnect us for stalling block download.
```

TheBlueMatt committed:
[qa] Make compact blocks test construction using fetch methods

TheBlueMatt commented:
Pushed fixes for @sdaftuar's comments, note that travis should fail the second-to-last-one as I added a test for the bug identified.

TheBlueMatt commented:
Rebased to clean up merge conflict, squashing the fixup commits. No changes to code.

TheBlueMatt committed:
Relay compact block messages prior to full block connection

TheBlueMatt commented:
Added a check that we only ever fast-announce once per block height, see https://0bin.net/paste/f1-cJq66rGul-VBY#vUW-L7QZR4+e8u/vx1ZNk1x3+5qn50ysFOghGC1gyLy
I dont think this is strictly required, but it simplifies review significantly and is likely a better behavior anyway, since you can never reconstruct two compact blocks at the same height reasonable right now anyway.

Also fixed a rebase issue where a diff chunk ended up in the wrong commit.

ryanofsky commented on src/net_processing.cpp:
```diff
@@ -2859,13 +2861,24 @@ bool SendMessages(CNode* pto, CConnman& connman, std::atomic<bool>& interruptMsg
                     int nSendFlags = state.fWantsCmpctWitness ? 0 : SERIALIZE_TRANSACTION_NO_WITNESS;
-                    connman.PushMessage(pto, msgMaker.Make(nSendFlags, NetMsgType::CMPCTBLOCK, cmpctblock));
+
+                    LOCK(cs_most_recent_block);
```
Might be good to release cs_most_recent_block in the else case during ReadBlockFromDisk.

TheBlueMatt committed:
Avoid holding cs_most_recent_block while calling ReadBlockFromDisk

morcos commented:
I'm the boy who cried ACK (4 times already) on this PR, so I'll refrain from repeating it again, but your last changes look good to me and I feel like all open concerns are addressed.

I'm going to put it into action combined with #9441 on a few nodes.

morcos commented:
@TheBlueMatt What do you think about making the LogPrints say:
`sending header-and-ids 0000myluckyhash to peer=8888`
instead of
`sending header-and-ids 0000myluckyhash to peer 8888`

You only add one of those in this PR, the other LogPrint for header-and-ids does the same thing, but its a bit annoying that they are formatted differently than other net debug messages.

TheBlueMatt commented:
@morcos did it separately in #9486.

laanwj commented on src/validationinterface.h:
```diff
@@ -8,6 +8,7 @@
 
 #include <boost/signals2/signal.hpp>
 #include <boost/shared_ptr.hpp>
+#include <memory>
```
This include is unnecessary

TheBlueMatt commented:
The <memory> include is only unnecessary because I'm sure boost/shared_ptr includes it. It is appropriate to add because we're now using shared_ptrs in one of the functions.

On January 10, 2017 4:47:55 AM PST, "Wladimir J. van der Laan" <notifications@github.com> wrote:
>laanwj commented on this pull request.
>
>
>
>> @@ -8,6 +8,7 @@
> 
> #include <boost/signals2/signal.hpp>
> #include <boost/shared_ptr.hpp>
>+#include <memory>
>
>This include is unnecessary
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/bitcoin/bitcoin/pull/9375#pullrequestreview-15907842

laanwj commented:
Ok, fair enough.

laanwj commented:
utACK c1ae4fc

theuni commented:
utACK all except https://github.com/bitcoin/bitcoin/pull/9375#discussion_r95670603. (see IMO we need a more general-purpose solution for fencing here, otherwise we're asking for future bugs.

TheBlueMatt committed:
Use a temp pindex to avoid a const_cast in ProcessNewBlockHeaders

TheBlueMatt commented on src/net_processing.cpp:
```diff
@@ -1558,17 +1597,7 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
-        }
-        int nSendFlags = State(pfrom->GetId())->fWantsCmpctWitness ? 0 : SERIALIZE_TRANSACTION_NO_WITNESS;
-        connman.PushMessage(pfrom, msgMaker.Make(nSendFlags, NetMsgType::BLOCKTXN, resp));
+        SendBlockTransactions(block, req, pfrom, connman);
```
We take cs_main again inside SendBlockTransactions, so I dont think its worth unlocking here.

sipa replied:
It probably doesn't matter too much here, but it's easier to reason about function that are only called with certain locks held or not held, and not both depending on the circumstances.

TheBlueMatt replied:
I hope to remove the cs_main in the function (and then not call it with cs_main) after #9419 + a few more changes.

TheBlueMatt committed:
Call ActivateBestChain without cs_main/with most_recent_block

There is still a call to ActivateBestChain with cs_main if a peer
requests the block prior to it being validated, but this one is
more specifically-gated, so should be less of an issue.

TheBlueMatt committed:
Add comment to describe callers to ActivateBestChain

TheBlueMatt commented:
Address all of @sipa's comments, and added a comment to (hopefully) appease @theuni.

morcos commented:
OK, I'm going for number 5!

tested ACK c1ae4fc
utACK 73666ad

To reemphasize, the caching in this PR provides a huge performance win.

theuni commented:
@TheBlueMatt Thanks for adding the comment.  utACK 73666ad05932429c860efe74eb388d212c152fc4.

sdaftuar commented:
utACK 73666ad

sipa commented on src/net_processing.cpp:
```diff
@@ -2859,13 +2861,24 @@ bool SendMessages(CNode* pto, CConnman& connman, std::atomic<bool>& interruptMsg
+                    LOCK(cs_most_recent_block);
+                    if (most_recent_block_hash == pBestIndex->GetBlockHash()) {
+                        if (state.fWantsCmpctWitness)
+                            connman.PushMessage(pto, msgMaker.Make(nSendFlags, NetMsgType::CMPCTBLOCK, *most_recent_compact_block));
```
Idea for follow-up PR: cache the `CSerializedNetMsg` instead of the `CBlockHeaderAndShortTxIDs` (smaller and faster to process)?

sipa commented:
utACK 73666ad05932429c860efe74eb388d212c152fc4

TheBlueMatt committed:
Make most_recent_compact_block a pointer to a const

TheBlueMatt commented on src/net_processing.cpp:
```diff
@@ -754,10 +754,11 @@ void PeerLogicValidation::SyncTransaction(const CTransaction& tx, const CBlockIn
 
 static CCriticalSection cs_most_recent_block;
 static std::shared_ptr<const CBlock> most_recent_block;
+static std::shared_ptr<CBlockHeaderAndShortTxIDs> most_recent_compact_block;
```
Ok, done.

TheBlueMatt commented:
Pushed one (hopefully last) commit which only adds two const's to net_processing at @sipa's request.

sipa commented on src/net_processing.cpp:
```diff
@@ -754,11 +754,11 @@ void PeerLogicValidation::SyncTransaction(const CTransaction& tx, const CBlockIn
 
 void PeerLogicValidation::NewPoWValidBlock(const CBlockIndex *pindex, const std::shared_ptr<const CBlock>& pblock) {
-    std::shared_ptr<CBlockHeaderAndShortTxIDs> pcmpctblock = std::make_shared<CBlockHeaderAndShortTxIDs> (*pblock, true);
+    std::shared_ptr<const CBlockHeaderAndShortTxIDs> pcmpctblock = std::make_shared<CBlockHeaderAndShortTxIDs> (*pblock, true);
```
Also add const to the make_shared parameter?

TheBlueMatt replied:
Ok, did so and squashed it into the previous commit.

rebroad commented on src/net_processing.cpp:
```diff
@@ -912,6 +957,21 @@ void static ProcessGetData(CNode* pfrom, const Consensus::Params& consensusParam
+                            a_recent_block = most_recent_block;
+                        }
+                        CValidationState dummy;
+                        ActivateBestChain(dummy, Params(), a_recent_block);
```
I'm struggling to understand why this is really needed, unless it is here to cater for a re-org that goes back more than 30 days as the logic below allows blocks to be sent that are not in the main chain but are less than 30 days old.
